### DISCUSSION TITLE
DataChange of Item in MapSchema include changes not correct

### DISCUSSION THREAD

**Discussion by vinkent420** [colyseus/colyseus#618] [Q&A] [answered]
Hi @endel and community developer!

I'm in progressing of upgrading to version 15 (I used latest version for both all client, server, schema, redis-driver... all are latest).
I the room state, I have a field players: MapSchema<IPlayer>.

At **client side**, I registered at 

```
this.room.state.player.onAdd(player, key) => {
  player.onChange(changes => {
    console.log(key, changes) // <--- changes here is array changes of ALL PLAYERS
  })
}
```
I used "changes" to have some specific logic, and it made me wrong :(. In the console.log I see changes includes of players (opposite with though it's on specific player (cause its callback in player.onChange)

Ah, when I log, I see refId, I search on document and don't see doc mention about,  can I using refId as a useful field to solve some business logic.

Thank @endel and Community.

**Comment by endel** [comment, marked as answer]
HI @vinkent420, I've just documented this missing piece in the migration guide here: https://docs.colyseus.io/migrating/0.15/#on-schema-instances

The `changes` list is now officially deprecated and won't exist anymore 
🙌  https://github.com/colyseus/schema/commit/7a99c3de5b6e4602ddf816d3dafe339d4f889238

You can use `.listen()` to act on specific properties, which is how most people used to do looping through each of the `changes` anyway!

Please let me know if you have any thoughts or concerns about this change. Cheers!

**Reply by vinkent420** [reply]
👍 Thank @endel so so much, you are right 🙌. After update my code follow your guide, bump, it worked well, haha.

Ah Endel, I must admit that in new version of 0.15 a lots of performance were improved, I feel that :-).
